clang/test/SemaSYCL/check-direct-attribute-propagation.cpp

```diff
@@ -0,0 +1,289 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -Wno-sycl-2017-compat -sycl-std=2020 -ast-dump %s | FileCheck %s
+
+// Tests for AST of [[intel::scheduler_target_fmax_mhz()]], [[intel::num_simd_work_items()]],
+// [[intel::no_global_work_offset()]], [[intel::no_global_work_offset()]], [[intel::sycl_explicit_simd]],
+// [[sycl::reqd_sub_group_size()]], [[sycl::reqd_work_group_size()]], [[intel::kernel_args_restrict]], and
+// [[intel::max_work_group_size()]] function attributes in SYCL 2020.
+
+#include "sycl.hpp"
+
+sycl::queue deviceQueue;
+
+struct FuncObj {
+  [[intel::sycl_explicit_simd]] void operator()() const {}
+};
+
+struct FuncObj1 {
+  [[intel::no_global_work_offset(1)]] void operator()() const {}
+};
+
+struct FuncObj2 {
+  [[intel::scheduler_target_fmax_mhz(10)]] void operator()() const {}
+};
+
+struct FuncObj3 {
+  [[intel::max_work_group_size(2, 2, 2)]] void operator()() const {}
+};
+
+struct FuncObj4 {
+  [[sycl::reqd_work_group_size(2, 2, 2)]] void operator()() const {}
+};
+
+struct FuncObj5 {
+  [[intel::num_simd_work_items(8)]] void operator()() const {}
+};
+
+struct FuncObj6 {
+  [[intel::kernel_args_restrict]] void operator()() const {}
+};
+
+struct FuncObj7 {
+  [[intel::max_global_work_dim(1)]] void operator()() const {}
+};
+
+[[intel::sycl_explicit_simd]] void func() {}
+
+[[intel::no_global_work_offset(1)]] void func1() {}
+
+[[intel::scheduler_target_fmax_mhz(2)]] void func2() {}
+
+[[intel::max_work_group_size(1, 1, 1)]] void func3() {}
+
+[[sycl::reqd_work_group_size(1, 1, 1)]] void func4() {}
+
+[[intel::num_simd_work_items(5)]] void func5() {}
+
+[[intel::kernel_args_restrict]] void func6() {}
+
+[[intel::max_global_work_dim(0)]] void func7() {}
+
+[[sycl::reqd_sub_group_size(4)]] void func8() {}
+
+class Functor {
+public:
+  void operator()() const {
+    func8();
+  }
+};
+
+class Functor1 {
+public:
+  [[sycl::reqd_sub_group_size(12)]] void operator()() const {}
+};
+
+int main() {
+  deviceQueue.submit([&](sycl::handler &h) {
+    // CHECK:       FunctionDecl {{.*}}test_kernel1
+    // CHECK:       SYCLSimdAttr {{.*}} Implicit
+    // CHECK-NEXT:  SYCLKernelAttr {{.*}} Implicit
+    // CHECK-NEXT:  SYCLSimdAttr
+    h.single_task<class test_kernel1>(
+        FuncObj());
+    // CHECK:       FunctionDecl {{.*}}test_kernel2
+    // CHECK:       SYCLSimdAttr {{.*}} Implicit
+    // CHECK-NEXT:  SYCLKernelAttr {{.*}} Implicit
+    // CHECK-NEXT:  SYCLSimdAttr
+    h.single_task<class test_kernel2>(
+        []() [[intel::sycl_explicit_simd]]{});
+
+    // Test attribute is not propagated.
```

Comment:
Can you please paste the AST for this test and the one on L80. I'm a little confused.

Reply:
> Can you please paste the AST for this test and the one on L80. I'm a little confused.

AST for 2020 - no propagation
```
FunctionDecl 0x560d39e7bf50 <line:96:9, <invalid sloc>> col:9 _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE_clES2_E12test_kernel3 'void ()'
| |-CompoundStmt 0x560d39e7c7c0 <<invalid sloc>>
| | |-DeclStmt 0x560d39e7c6c0 </export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp:231:23>
| | | `-VarDecl 0x560d39e7c618 </export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/check-direct-attribute-propagation.cpp:96:9, /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp:231:23> /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/check-direct-attribute-propagation.cpp:96:9 used '(lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/check-direct-attribute-propagation.cpp:96:9)' cinit
| | |   `-InitListExpr 0x560d39e7c680 </export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp:231:23> '(lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/check-direct-attribute-propagation.cpp:96:9)'
| | `-CompoundStmt 0x560d39e7c7a8 <col:72, line:233:1>
| |   `-CXXOperatorCallExpr 0x560d39e7c778 <line:232:3, col:14> 'void':'void' '()'
| |     |-ImplicitCastExpr 0x560d39e7c748 <col:13, col:14> 'void (*)() const' <FunctionToPointerDecay>
| |     | `-DeclRefExpr 0x560d39e7c728 <col:13, col:14> 'void () const' lvalue CXXMethod 0x560d39e57fe0 'operator()' 'void () const'
| |     `-ImplicitCastExpr 0x560d39e7c760 <col:3> 'const (lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/check-direct-attribute-propagation.cpp:96:9)' lvalue <NoOp>
| |       `-DeclRefExpr 0x560d39e7c708 <col:3> '(lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/check-direct-attribute-propagation.cpp:96:9)' lvalue Var 0x560d39e7c618 '' '(lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/check-direct-attribute-propagation.cpp:96:9)'
| |-OpenCLKernelAttr 0x560d39e7bff0 <<invalid sloc>> Implicit
| |-AsmLabelAttr 0x560d39e7c550 <<invalid sloc>> Implicit "_ZTSZZ4mainENKUlRN2cl4sycl7handlerEE_clES2_E12test_kernel3"
| |-ArtificialAttr 0x560d39e7c5c8 <<invalid sloc>> Implicit
| |-SYCLSimdAttr 0x560d39e7c5f0 <<invalid sloc>> Implicit
| |-SYCLKernelAttr 0x560d39e7c7e0 <<invalid sloc>> Implicit

```

AST- 2017 mode with propagation (sycl-esimd-ast.cpp)
```
FunctionDecl 0x561f7fbd92e0 <line:34:9, <invalid sloc>> col:9 _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE_clES2_E12test_kernel3 'void ()'
  |-CompoundStmt 0x561f7fbd9648 <<invalid sloc>>
  | |-DeclStmt 0x561f7fbd9548 </export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp:231:23>
  | | `-VarDecl 0x561f7fbd94a0 </export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/sycl-esimd-ast.cpp:34:9, /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp:231:23> /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/sycl-esimd-ast.cpp:34:9 used '(lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/sycl-esimd-ast.cpp:34:9)' cinit
  | |   `-InitListExpr 0x561f7fbd9508 </export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp:231:23> '(lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/sycl-esimd-ast.cpp:34:9)'
  | `-CompoundStmt 0x561f7fbd9630 <col:72, line:233:1>
  |   `-CXXOperatorCallExpr 0x561f7fbd9600 <line:232:3, col:14> 'void':'void' '()'
  |     |-ImplicitCastExpr 0x561f7fbd95d0 <col:13, col:14> 'void (*)() const' <FunctionToPointerDecay>
  |     | `-DeclRefExpr 0x561f7fbd95b0 <col:13, col:14> 'void () const' lvalue CXXMethod 0x561f7fbd1730 'operator()' 'void () const'
  |     `-ImplicitCastExpr 0x561f7fbd95e8 <col:3> 'const (lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/sycl-esimd-ast.cpp:34:9)' lvalue <NoOp>
  |       `-DeclRefExpr 0x561f7fbd9590 <col:3> '(lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/sycl-esimd-ast.cpp:34:9)' lvalue Var 0x561f7fbd94a0 '' '(lambda at /export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/sycl-esimd-ast.cpp:34:9)'
  |-OpenCLKernelAttr 0x561f7fbd9380 <<invalid sloc>> Implicit
  |-AsmLabelAttr 0x561f7fbd93d8 <<invalid sloc>> Implicit "_ZTSZZ4mainENKUlRN2cl4sycl7handlerEE_clES2_E12test_kernel3"
  |-ArtificialAttr 0x561f7fbd9450 <<invalid sloc>> Implicit
  |-SYCLSimdAttr 0x561f7fbd9478 <<invalid sloc>> Implicit
  |-SYCLKernelAttr 0x561f7fbd9668 <<invalid sloc>> Implicit
  |-SYCLSimdAttr 0x561f7fbd17e0 </export/iusers/smanna/my_sycl_ws/llvm/clang/test/SemaSYCL/sycl-esimd-ast.cpp:34:16, col:23>
  `-SYCLSimdAttr 0x561f7fbcbf48 <line:13:3, col:10>

```

Reply:
ok Thank you. I didn't notice the `implicit` and was wondering why we have so many nodes.  I wonder if we actually need to add duplicate attributes (not just SYCLSimdAttr. I mean all) when we collect and propagate attributes. If we don't, we should probably look into that. Anyway, that is outside the scope of this PR . Thanks again for explaining!

Reply:
> ok Thank you. I didn't notice the `implicit` and was wondering why we have so many nodes. I wonder if we actually need to add duplicate attributes (not just SYCLSimdAttr. I mean all) when we collect and propagate attributes. If we don't, we should probably look into that. Anyway, that is outside the scope of this PR . Thanks again for explaining!

@elizabethandrews, thanks for pointing this. I agree. I have created https://github.com/intel/llvm/issues/4119 for future reference. I will take a look at this later since this is outside the scope of this PR.